packages/@aws-cdk/aws-kinesisfirehose/README.md

```diff
@@ -9,8 +9,216 @@
+permissions for Kinesis Data Firehose to access the resources referenced by your delivery
+stream. For example: an Elasticsearch domain, a Redshift cluster, a backup or destination
+S3 bucket, a Lambda data transformer, an AWS Glue table schema, etc. If you wish, you may
+specify your own IAM role. It must have the correct permissions, or delivery stream
+creation or data delivery may fail.
+
+```ts fixture=with-bucket
+import * as iam from '@aws-cdk/aws-iam';
+
+const role = new iam.Role(this, 'Role', {
+  assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
+}
+bucket.grantWrite(role);
+new DeliveryStream(stack, 'Delivery Stream', {
+  destinations: [new destinations.S3(bucket)],
+  role: role,
+});
+```
+
+See [Controlling Access](https://docs.aws.amazon.com/firehose/latest/dev/controlling-access.html)
+in the *Kinesis Data Firehose Developer Guide*.
+
+## Granting application access to a delivery stream
+
+IAM roles, users or groups which need to be able to work with delivery streams should be
+granted IAM permissions.
+
+Any object that implements the `IGrantable` interface (ie., has an associated principal)
+can be granted permissions to a delivery stream by calling:
+
+- `grantPutRecords(principal)` - grants the principal the ability to put records onto the
+  delivery stream
+- `grant(principal, ...actions)` - grants the principal permission to a custom set of
+  actions
+
+```ts fixture=with-delivery-stream
+import * as iam from '@aws-cdk/aws-iam';
+const lambdaRole = new iam.Role(this, 'Role', {
+  assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
+}
+
+// Give the role permissions to write data to the delivery stream
+deliveryStream.grantPutRecords(lambdaRole);
+```
+
+The following write permissions are provided to a service principal by the `grantPutRecords()` method:
+
+- `firehose:PutRecord`
+- `firehose:PutRecordBatch`
+
+## Granting a delivery stream access to a resource
+
+Conversely to the above, Kinesis Data Firehose requires permissions in order for delivery
+streams to interact with resources that you own. For example, if an S3 bucket is specified
+as a destination of a delivery stream, the delivery stream must be granted permissions to
+put and get objects from the bucket. When using the built-in AWS service destinations
+found in the `@aws-cdk/aws-kinesisfirehose-destinations` module, the CDK grants the
+permissions automatically. However, custom or third-party destinations may require custom
+permissions. In this case, use the delivery stream as an `IGrantable`, as follows:
+
+```ts fixture=with-delivery-stream
+/// !hide
+const myDestinationResource = {
+  grantWrite(grantee: IGrantable) {}
+}
+/// !show
+myDestinationResource.grantWrite(deliveryStream);
+```
+
+## Multiple destinations
+
+Though the delivery stream allows specifying an array of destinations, only one
+destination per delivery stream is currently allowed. This limitation is enforced at
+compile time and will throw an error.
```

Comment:
I would also remove the description of who checks the error.

Reply:
I decided to leave the validation in, because the current implementation will not just work as-is if the service starts to support multiple destinations.  We don't know what the API will be if/when that feature is available. Will it just be you can supply both ExtendedS3DestinationConfiguration and RedshiftDestinationConfiguration? Or will those fields start to accepts arrays? We don't know, so some work will be required in the CDK anyway.

Reply:
Sure. Makes sense.

Let's just substitute "compile time" for "CDK synthesis time", to be 100% precise.